src/subcommand/server/rpc.rs

```diff
@@ -121,6 +153,106 @@ async fn get_sat_ranges(value: JsonRpcExtractor, index: Arc<Index>) -> JrpcResul
   Ok(JsonRpcResponse::success(answer_id, res))
 }
 
+fn get_block_rarities(start: u64, end: u64) -> Result<BlockRarities> {
+  let mut vintage = 0;
+  let mut nakamoto = 0;
+  let mut first_transaction = 0;
+  let mut pizza = 0;
+  let mut block9 = 0;
+  let mut block78 = 0;
+  let palindrome;
+
+  let sat_start = Sat(start);
+  let sat_end = Sat(end);
+
+  if start > end {
+    return Err(anyhow!("invalid sat range: start {start} > end {end}"));
+  }
+  if sat_start.height().n() != sat_end.height().n() {
+    return Err(anyhow!("invalid sat range: start {start} and end {end} are in different blocks"));
+  }
+
+  let block_height = sat_start.height().n();
+
+  // ignore sat ranges later than the max pizza block
+  if block_height <= MAX_PIZZA_BLOCK_HEIGHT {
+    if block_height <= VINTAGE_BLOCK_HEIGHT {
+      vintage += sat_end.n() - sat_start.n() + 1;
+    }
+      
+    if NAKAMOTO_BLOCK_HEIGHTS.contains(&block_height) {
+      nakamoto += sat_end.n() - sat_start.n() + 1;
+    }
+
+    if block_height == BLOCK9_BLOCK_HEIGHT {
+      block9 = sat_end.n() - sat_start.n() + 1;
+      if sat_start.n() < FIRST_TRANSACTION_SAT_RANGE.1 {
+        first_transaction = min(FIRST_TRANSACTION_SAT_RANGE.1 - 1, sat_end.n()) - max(FIRST_TRANSACTION_SAT_RANGE.0, sat_start.n()) +1;
+      }
+    } else if block_height == BLOCK78_BLOCK_HEIGHT {
+      block78 = sat_end.n() - sat_start.n() + 1;
+    }
+
+    if PIZZA_RANGE_MAP.contains_key(&block_height) {
+      let pizza_sat_ranges = PIZZA_RANGE_MAP.get(&block_height).unwrap();
+      pizza += count_pizza_sats_in_range(start, end, &pizza_sat_ranges);
+    }
+  }
+
+  // Assume sat ranges are not too large for most of the cases.
```

Comment:
question: if we need to calculate the palindrome in the BE side anyway, since we can't fully rely on the estimated result, do we still need to calculate here?